File: Exercice3/app/src/main/java/es/ua/eps/exercice3/MainActivity.java
package es.ua.eps.exercice3;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.Settings;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.util.Objects;

import es.ua.eps.exercice3.databinding.ActivityMainBinding;
import es.ua.eps.exercice3.databinding.ActivityShowInfoBinding;

public class MainActivity extends AppCompatActivity {

    final String INFO_TYPE = "INFO_TYPE";
    final String LOCATION_TYPE = "LOCATION_TYPE";     //if 0 local if 1 external

    ActivityMainBinding viewBinding;

    TextView infoTextView;
    EditText textToSave;

    Button seeExternalstateButton;
    Button saveText;
    Button seeFileButton;

    Button moveToExternal;
    Button moveToInternal;

    Button closeButton;


    boolean sdAvaible = false;
    boolean sdWriteAccess = false;
    boolean checkExternalStorage = false;

    int locationType = 0;

    String filePath = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        viewBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(viewBinding.getRoot());


        infoTextView = viewBinding.moveToText;

        textToSave = viewBinding.textToSave;
        seeExternalstateButton = viewBinding.seeExternalStorage;
        saveText = viewBinding.addFile;
        seeFileButton = viewBinding.seeFile;
        closeButton = viewBinding.closeButton;

        moveToInternal = viewBinding.internalStorageButton;
        moveToExternal = viewBinding.externalStorageButton;

        moveToInternal.setEnabled(false);
        seeExternalstateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                Intent intent = new Intent(view.getContext(), ShowInfoActivity.class);
                intent.putExtra(INFO_TYPE, 0);
                startActivity(intent);
            }
        });

        saveText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                SaveText();
            }
        });

        seeFileButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                Intent intent = new Intent(view.getContext(), ShowInfoActivity.class);
                intent.putExtra(INFO_TYPE, 1);
                intent.putExtra(LOCATION_TYPE, locationType);
                startActivity(intent);
            }
        });


        closeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });

        moveToInternal.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ChangeToInternal();
            }
        });

        moveToExternal.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (permisions()) {
                    ChangeToExternal();
                }
            }
        });

    }

    private Boolean existFile(String[] files, String fileName) {
        for (String file : files) {
            if (Objects.equals(file, fileName)) {
                return true;
            }
        }
        return false;
    }


    private void SaveText() {

        String state = Environment.getExternalStorageState();
        if (state.equals(Environment.MEDIA_MOUNTED)) {
            sdAvaible = true;
            sdWriteAccess = true;
        } else if (state.equals(Environment.MEDIA_MOUNTED_READ_ONLY)) {
            sdAvaible = true;
            sdWriteAccess = false;
        }

        try {

            File sdPath = null;
            if (locationType == 0)
            {
                sdPath = getFilesDir();
            }
            else {
                sdPath = Environment.getExternalStorageDirectory();
            }

            File f = new File(sdPath, "information.txt");

            OutputStreamWriter fout = new OutputStreamWriter(new FileOutputStream(f));
            fout.write(textToSave.getText().toString());
            fout.flush();
            fout.close();

        } catch (Exception e) {

        }
    }

    private Boolean permisions() {
        if (Build.VERSION.SDK_INT >= 30) {
            if (Environment.isExternalStorageManager() == false) {
                Uri uri = Uri.parse("package:" + BuildConfig.APPLICATION_ID);
                startActivity(new Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION, uri));
                checkExternalStorage = true;
                return false;
            } else {
                return true;
            }
        }
        return true;
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        if (Build.VERSION.SDK_INT >= 30) {
            if(checkExternalStorage) {
                checkExternalStorage = false;
                if(Environment.isExternalStorageManager()) {
                   ChangeToExternal();
                }
            }
        }
    }

    void MoveFromLocalToExternal(){MoveFile(getFilesDir(), Environment.getExternalStorageDirectory());}
    void MoveFromExternalToLocal(){MoveFile(Environment.getExternalStorageDirectory(), getFilesDir());}

    void MoveFile(File originDir, File destinyDir){
        try{
            File originFile = new File(originDir,"information.txt" );
            File destinyFile = new File (destinyDir, "information.txt");

            FileInputStream fileInputStream = new FileInputStream(originFile);
            FileOutputStream fileOutputStream = new FileOutputStream(destinyFile);

            byte [] buffer = new byte [1024];
            int length = fileInputStream.read(buffer);
            while (length > 0){
                fileOutputStream.write(buffer, 0, length);
                length = fileInputStream.read(buffer);
            }

            fileInputStream.close();
            fileOutputStream.close();

            originFile.delete();



        }catch(Exception e)
        {
            e.printStackTrace();
        }
    }

    private void ChangeToExternal()
    {
        locationType = 1;
        MoveFromLocalToExternal();
        moveToInternal.setEnabled(true);
        moveToExternal.setEnabled(false);
    }

    private void ChangeToInternal()
    {
        locationType = 0;
        MoveFromExternalToLocal();
        moveToInternal.setEnabled(false);
        moveToExternal.setEnabled(true);
    }

}

File: Exercice4/app/src/main/java/es/ua/eps/exercice4/MainActivity.kt
package es.ua.eps.exercice4

import android.content.ContentValues
import android.content.Intent
import android.net.Uri
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Environment
import android.provider.Settings
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.core.app.NavUtils
import es.ua.eps.exercice4.databinding.ActivityMainBinding
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

class MainActivity : AppCompatActivity() {

    lateinit var viewBinding: ActivityMainBinding

    lateinit var usernameEditText : EditText
    lateinit var passwordEditText : EditText

    lateinit var loginButton: Button
    lateinit var closeButton: Button

    private lateinit var sqliteHelper : UsersSQLiteHelper

    companion object {
        const val USER_NAME = "USER_NAME"
        const val USER_COMPLETE_NAME = "USER_COMPLETE_NAME"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)

        sqliteHelper = UsersSQLiteHelper(this)

        supportActionBar?.title = "SQLite"

        usernameEditText = viewBinding.userNameEditText
        passwordEditText = viewBinding.passwordEditText

        loginButton = viewBinding.loginButton

        loginButton.setOnClickListener {
            val us = usernameEditText.text.toString()
            val password = passwordEditText.text.toString()
            var loginSuccessful = false
            if(us.isNotEmpty() && password.isNotEmpty()){
                val users = getUsers()
                println("users -> ${users.size}")

                for (user in users) {
                    val username = user.getAsString(UserContentProvider.username)
                    val storedPassword = user.getAsString(UserContentProvider.password)
                    val userCompleteName = user.getAsString(UserContentProvider.nombreCompleto)

                    println("us ${us} -> $username")
                    println("pas $password -> $storedPassword")
                    if (us.equals(username) && password.equals(storedPassword)) {
                        val intentUser = Intent(this@MainActivity, UserDataActivity::class.java)
                        intentUser.putExtra(USER_NAME, username)
                        intentUser.putExtra(USER_COMPLETE_NAME, userCompleteName)
                        startActivity(intentUser)
                        loginSuccessful = true
                        break
                    }
                }
                if (!loginSuccessful) {
                    Toast.makeText(this, "Error in data", Toast.LENGTH_SHORT).show()
                }
            }else{
                Toast.makeText(this,"Error blank fields", Toast.LENGTH_SHORT).show()
            }
        }

        closeButton = viewBinding.closeButton
        closeButton.setOnClickListener{
            finish();
        }

    }


private fun getUsers(): ArrayList<ContentValues>{
    val cursor = contentResolver.query(
        UserContentProvider.CONTENT_URI,
        arrayOf(UserContentProvider.id, UserContentProvider.username, UserContentProvider.password, UserContentProvider.nombreCompleto, UserContentProvider.email),
        null,
        null,
        UserContentProvider.id
    )

    val users = ArrayList<ContentValues>()

    cursor?.let {

        val idColumnIndex = it.getColumnIndex(UserContentProvider.id)
        val usernameColumnIndex = it.getColumnIndex(UserContentProvider.username)
        val passwordColumnIndex = it.getColumnIndex(UserContentProvider.password)
        val completenameColumnIndex = it.getColumnIndex(UserContentProvider.nombreCompleto)

        while (it.moveToNext()) {
            println("Dent")

            println("idIndx $idColumnIndex idUserName $usernameColumnIndex id pass $passwordColumnIndex idcomplete" +
                    " $completenameColumnIndex" )


            // Verificar que los índices sean válidos
            if (idColumnIndex >= 0 && usernameColumnIndex >= 0 &&
                passwordColumnIndex >= 0 && completenameColumnIndex >= 0
            ) {
                println("Harvey")

                // Obtener los datos usando los índices
                val cv = ContentValues()
                cv.put(UserContentProvider.id, it.getInt(idColumnIndex))
                cv.put(UserContentProvider.username, it.getString(usernameColumnIndex))
                cv.put(UserContentProvider.password, it.getString(passwordColumnIndex))
                cv.put(UserContentProvider.nombreCompleto, it.getString(completenameColumnIndex))
                users.add(cv)
            }
        }
    }
    cursor?.close()
    return users

}

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id: Int = item.getItemId()

        when(id)
        {

            R.id.manageUsers ->  {
                val intent = Intent(this, UserManagementActivity::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
                return true
            }
            R.id.createBU -> BackUpManager.createBackup(
                this,
                Intent(
                    Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION,
                    Uri.parse("package:" + BuildConfig.APPLICATION_ID)))
            R.id.restoreBU -> BackUpManager.restoreBackup(this)
        }


        return super.onOptionsItemSelected(item)
    }

    override fun onRestart() {
        super.onRestart()
        if (Build.VERSION.SDK_INT >= 30) {
            if (BackUpManager.checkExternalStorage) {
                if (Environment.isExternalStorageManager()) {
                    //CREATE BACKUP
                    BackUpManager.backupSave(this)
                }
            }
        }
    }


}